build-system/npm-publish/utils.js

```diff
@@ -70,11 +74,126 @@ function getSemver(extensionVersion, ampVersion) {
   return `${major}.${minor}.${patch}`;
 }
 
+/**
+ * @typedef {{
+ *  name: string,
+ *  version: string,
+ *  author: string,
+ *  license: string,
+ *  main: string,
+ *  module: string,
+ *  description: string,
+ *  homepage: string,
+ *  files: string[],
+ *  repostitory: {
+ *   type: string,
+ *   url: string,
+ *   directory: string
+ *  },
+ *  peerDependencies: Record<string, string>,
+ *  dependencies: Record<string, string>,
+ * }}
+ */
+let PackageFileDef;
+
+/**
+ * Get the parsed package files
+ * @return {Promise<PackageFileDef[]>}
+ */
+async function getPackageFiles() {
+  const [extPackageFiles, componentPackageFiles] = await Promise.all([
+    fastGlob(path.join('extensions', '**', 'package.json')),
+    fastGlob(path.join('src', 'bento', 'components', '**', 'package.json')),
+  ]);
+
+  return await Promise.all(
+    [...extPackageFiles, ...componentPackageFiles].map(async (filePath) => {
+      const file = await readFile(filePath, 'utf-8');
+      return JSON.parse(file);
+    })
+  );
+}
+
+/**
+ * Build a graph of package names to packages (including their dependencies)
+ * @param {PackageFileDef[]} packageFiles
+ * @return {Record<string, PackageFileDef>}
+ */
+function buildPublishGraph(packageFiles) {
+  return packageFiles.reduce((graph, file) => {
+    graph[file.name] = file;
+    return graph;
+  }, {});
+}
+
+/**
+ *
+ * @param {Record<string, PackageFileDef>} graph
+ * @return {string[]}
+ */
+function getPackagesWithoutDepeneceiesInGraph(graph) {
+  return Object.keys(graph).filter((name) => {
+    const node = graph[name];
+    if (!node.dependencies) {
+      return true;
+    }
+    const dependenciesInGraph = Object.keys(node?.dependencies).filter(
+      (dep) => graph[dep]
+    );
+    return !Boolean(dependenciesInGraph.length);
+  });
+}
+
+/**
+ * Because our packages rely on specific versions of each other, dependencies need to
+ * be published before their dependents. This function solves the dependency graph
+ * and returns the order the packages should be published in.
+ * @param {Record<string, PackageFileDef>} graph
+ * @return {PackageFileDef[]}
+ */
+function solveDependencyGraph(graph) {
```

Comment:
can we call this `getTopologicalSort` or something similar since that's what it's doing?